### DISCUSSION TITLE
extract fingerprints for SMARTS?

### DISCUSSION THREAD

**Discussion by chaoyan1037** [rdkit/rdkit#4513] [General]
Is there a way to extract fingerprints for SMARTS? Converting them into SMILES may result in invalid molecules. PatternFingerprint does not consider the bond information. I am looking for something else than can be used as the representation for the SMARTS sequences.

For example, I have the following examples:

```
'C-[C&H1&D2&+0]=[C&H1&D2&+0]-[C&D1&H3]',
'C-[C&H1&D2&+0]=[C&H1&D2&+0]-C',
'C/[C&H1&D2&+0]=[C&H1&D2&+0]\\C',
'C/[C&H1&D2&+0]=[C&H1&D2&+0]\\[C&D1&H3]',
'C/C=C/[C&H3&D1&+0]',
'C/[C&H1&D2&+0]=[C&H1&D2&+0]/C'
```

**Comment by greglandrum** [comment]
@chaoyan1037 : are you asking if it's possible to create fingerprints for molecules which have been constructed from SMARTS?

If so: the RDKit does not have anything which would take query features on either atoms or bonds into account.

**Reply by chaoyan1037** [reply]
Can I ask why not?

**Reply by greglandrum** [reply]
It's never been something which I personally needed and I guess that must be true for other RDKit contributors as well.

There is an interesting intellectual problem here to figure out what could be used for the atom and bond invariants in order to give similarities which make sense. And, I guess, another interesting problem to figure out what exactly "makes sense" means in this context.

As a simple example of the complexity, consider these pairs:
```
[C]-[N]   [C]-[NH]
[C]-[N]   [C]-[N!H0]
[C]-[N]   [C]-[NH1]
[C]-[N]   [C]-[NH2]
[C]-[N]   [C]-[NH0]
[C]-[NH]   [C]-[NH2]
[C]-[NH1]   [C]-[NH2]
[C]-[NH0]   [C]-[NH1]
[C]-[N]    [C]-[#7]
```

What should the *relative* similarities of those pairs be?

**Reply by chaoyan1037** [reply]
Thanks for your explanation! My intention is not to compare the similarity, but to obtain a representation or embedding for each SMARTS  that can be used for machine learning models. If rdkit does not provide something like that, I can try other methods.
